### DISCUSSION TITLE
How do I learn real graphs (skeletons) from training on approximate (KNNGraph) graphs?

### DISCUSSION THREAD

**Discussion by davodogster** [pyg-team/pytorch_geometric#4924] [Q&A]
Hi guys,

This is my PyG dataset format:

![image](https://user-images.githubusercontent.com/46079516/178189652-75517af7-b2a1-4918-8517-fc3ca1c4a2e3.png)

**Comment by rusty1s** [comment]
Can you share some insights on what you want to do concretely and how X and Y relate to each other (since their number of nodes vary)?

**Reply by davodogster** [reply]
Hi @rusty1s Yes the number of nodes vary. Because Y is a ground truth graph (edges: X_edge_index) of a tree skeleton (plus vertices: X_pos), and X is a KNN graph (Y_edge_index) from a point cloud (Y_pos) of a tree. I hope that makes sense!

**Reply by rusty1s** [reply]
Thanks. With this, can you clarify what are you trying to do? Map nodes in `X` to nodes in `Y`? What's the task you are trying to solve?

**Reply by davodogster** [reply]
> Thanks. With this, can you clarify what are you trying to do? Map nodes in `X` to nodes in `Y`? What's the task you are trying to solve?

Hi @rusty1s Ideally, I would like to map nodes in X to nodes in Y and also the same for pos. Multitask learning. But worse case scenario I want to map X_edge_index to Y_edge_index. The task I am trying to solve is point cloud skeletonization and I have all the necessary data to train the model but I need some help in how to configure the model architecture.

**Reply by rusty1s** [reply]
Does it make sense to map edges (since they are synthetically created)? If you have the mapping from nodes, you already have the mapping to edges, right? I think you can train this via supervision, e.g., in case you no correspondences between `X` and `Y`, you can obtain a `[num_X_nodes, num_Y_nodes]` matrix based on the inner product of the two point cloud representations of a siamese network, and train it against your ground-truth.

**Reply by davodogster** [reply]
Hi @rusty1s this is a real world problem and I have the data to train it. If you collect a point cloud from a Tree in the real world, then you use deep learning to obtain (an estimate of the ground truth) Tree Graph (nodes, edges) / Skeleton. Yes I do have the ground truth Graph/edge_index tensor to use for training data. Does that make sense?? Cheers, Sam

**Reply by rusty1s** [reply]
Yes, this makes sense. Let me know if you have trouble to implement this :)

**Reply by davodogster** [reply]
I think I will need help! So i'll be in touch! Where is a good place to start? It's a bit like segmentation but different number of outputs to inputs?

**Reply by rusty1s** [reply]
Have a look at:
* https://github.com/pyg-team/pytorch_geometric/blob/master/examples/pointnet2_segmentation.py
* https://github.com/pyg-team/pytorch_geometric/blob/master/examples/faust.py

**Reply by davodogster** [reply]
Thanks, Hmm what is the edge_attr in the FAUST dataset?

**Reply by rusty1s** [reply]
It's the relative spatial distance between two nodes, e.g., `pos[j] - pos[i]`.

**Reply by davodogster** [reply]
Thanks, is there a PyG function to compute it and add it to the Data object ?

**Reply by rusty1s** [reply]
Have a look at `transform.Cartesian`.

**Reply by davodogster** [reply]
> Hi @rusty1s Yes the number of nodes vary. Because Y is a ground truth graph (edges: X_edge_index) of a tree skeleton (plus vertices: X_pos), and X is a KNN graph (Y_edge_index) from a point cloud (Y_pos) of a tree. I hope that makes sense!

Hi @rusty1s sorry I noticed typos for X and Ys. It should read like this:
 Hi @rusty1s Yes the number of nodes vary. Because Y is a ground truth graph (edges: Y_edge_index) of a tree skeleton (plus vertices: Y_pos), and X is a KNN graph (X_edge_index) from a point cloud (X_pos) of a tree. I hope that makes sense!

Let me know if it changes the task!
I'll look into transform.Cartesian,  Sam

**Reply by davodogster** [reply]
> Have a look at:
> 
> * https://github.com/pyg-team/pytorch_geometric/blob/master/examples/pointnet2_segmentation.py
> * https://github.com/pyg-team/pytorch_geometric/blob/master/examples/faust.py

Can a PyG network be constructed to have output shape different to the input shape? (different size in the second dimension). E.g:
![image](https://user-images.githubusercontent.com/46079516/178187488-1c09616d-7366-4c1c-ab12-b8b34ab4512b.png)

.. and two inputs and outputs per sample. Faust and pointnet seg appear to only return values for one input param and not multiple.
Thanks, Sam

**Reply by rusty1s** [reply]
Yes, you can map to any feature dimension per node. The more challenging problem is how we can effectively reduce the number of nodes such that `X_pos.size(0) == Y_pos.size(0)`.

**Reply by davodogster** [reply]
Hi @rusty1s 

My data looks likes this. 
I have 1. PCD (input) 2. Skeleton(output). 3. Mesh (green) For 10000+ trees.

![image](https://user-images.githubusercontent.com/46079516/194463271-148b816c-1073-44b6-b037-a78c7ebd7fc7.png)
![image](https://user-images.githubusercontent.com/46079516/194463136-f26443eb-0136-4205-b49e-15d52c63e0b4.png)
![image](https://user-images.githubusercontent.com/46079516/194464246-9e7dce06-83c7-435f-a70a-1db193e0f5ca.png)

![image](https://user-images.githubusercontent.com/46079516/194463316-ee301fde-185a-41fa-be1d-58fed81d18e0.png)

(Note: these three files are not the same sample sorry)

I'd like to train a GNN to learn skeletons from point clouds. As above - skeletons have different number of vertices/pos than the number of points in the point cloud. Do you know if it's possible to train such a network in PyG. Does X_pos.shape[0] have to equal Y_pos.shape[0] and X_edge_index.shape[1] have to equal Y_edge_index.shape[1]. Its a very useful real world application if we can get this to work. Thanks, Sam

**Reply by rusty1s** [reply]
There does not exist such kind of example in PyG yet, I am sorry. Do you know of any papers that tackle this kind of problem?

**Reply by davodogster** [reply]
Hi @rusty1s thanks for your response. This one appears closest: https://onlinelibrary.wiley.com/doi/epdf/10.1111/cgf.14151
This is also similar: https://github.com/clinplayer/Point2Skeleton

This is related but for mesh only (ignore the skeleton) https://github.com/weixk2015/DHSP3D

Do you think it's possible to implement in PyG?

**Reply by davodogster** [reply]
Hi @rusty1s do you think the PCD to mesh or PCD to skeleton is possible with GNN deep learning? e.g https://github.com/clinplayer/Point2Skeleton 

Thanks

**Reply by rusty1s** [reply]
Sorry for the late reply here. Point2Skeleton is a cool paper :) Wouldn't this be something that is directly applicable in your case? The paper uses a GNN (i.e. PointNet++) to learn the skeletal points already.
